### DISCUSSION TITLE
Cluster expected replication behaviour with `minSyncReplicas`/`maxSyncReplicas`

### DISCUSSION THREAD

**Discussion by thomasboussekey** [cloudnative-pg/cloudnative-pg#1277] [Q&A] [answered]
Hello,

I've tried to implement this [kind of cluster configuration](https://github.com/cloudnative-pg/cloudnative-pg/blob/main/tests/e2e/fixtures/sync_replicas/cluster-pgstatstatements.yaml.template) on a test environment:
* 3 instance replication
* minSyncReplicas = maxSyncReplicas = 1

Once deployed, I get this instance status:

```
Instances status
Name              Database Size  Current LSN  Replication role  Status  QoS        Manager Version  Node
----              -------------  -----------  ----------------  ------  ---        ---------------  ----
cnpg1-proto-db-1  50 MB          0/9B00B808   Primary           OK      Burstable  1.18.0           gke-gke-alpha-01-europe--default-653c-0af18f8e-0d6j
cnpg1-proto-db-3  50 MB          0/9B00B880   Standby (sync)    OK      Burstable  1.18.0           gke-gke-alpha-01-europe--default-653c-7de2f8a2-0y90
cnpg1-proto-db-2  50 MB          0/9B00B808   Standby (sync)    OK      Burstable  1.18.0           gke-gke-alpha-01-europe--default-653c-676a7eab-r444
```

I thought I would have this kind of configuration:

```
Instances status
Name              Database Size  Current LSN  Replication role
----              -------------  -----------  ----------------
cnpg1-proto-db-1  50 MB          0/9B00B808   Primary         
cnpg1-proto-db-3  50 MB          0/9B00B880   Standby (async)  
cnpg1-proto-db-2  50 MB          0/9B00B808   Standby (sync)  
```

* 1 standby instance would be in `sync` replication, in order to comply with the configuration
* 1 standby instance would be in `async` replication, because **Max synchronous replica** is set to `1` anbd we have 2 standbys.

My configuration seems to be correctly set on the cluster:

```console
$ kubectl -n "${NAMESPACE}" describe cluster | grep Replicas
        f:maxSyncReplicas:
        f:minSyncReplicas:
  Max Sync Replicas:        1
  Min Sync Replicas:        1
```

I tried to load activity on the instance and both standbys remained in `sync` mode during the test.

Is it the awaited behaviour?

**Comment by sxd** [comment, marked as answer]
Hi @thomasboussekey 

Yes, it's the expected behavior since we implement [Multiple Synchronous Standbys](https://www.postgresql.org/docs/current/warm-standby.html#SYNCHRONOUS-REPLICATION) and we use `ANY OF` which basically means `WAL records are replicated to at least the requested number of synchronous standbys in the list.` that means that it's required 1 but most of the time you will have 2 synchronous replicas, if you use the cnpg plugin to run pgbench inside the cluster you will see how at least one it's falling behind.

I'm converting this into a discussion for now

Regards

**Reply by thomasboussekey** [reply]
Hello @sxd,

Thanks for your explanation!

It's very clear :+1:

**Reply by boedy** [reply]
@sxd To me this explains how `minSyncReplicas` is handled. But I don't seem to understand what `maxSyncReplicas` tries to solve here.
